File: sort/quick.js
const quickSortHelper = (nums, start, end) => {
    if(start >= end) return
    let i = start
    let j = end
    let base = nums[end]
    while(i < j) {
      while(i < j && nums[i] <= base) {
        i++
      }
      nums[j] = nums[i]
      while (i < j && nums[j] >= base) {
        j--
      }
      nums[i] = nums[j]
    }
    nums[i] = base
    quickSortHelper(nums, start, i - 1)
    quickSortHelper(nums, j + 1, end)
  }
  
  const quickSort = nums => {
    quickSortHelper(nums, 0, nums.length - 1)
  }
  
  const nums = [49, 38, 65, 97, 76, 13, 27, 49]
  
  quickSort(nums)
  console.log(nums)

File: sort/quick.go
package sort

import "fmt"

func quickSortHelper(nums []int, start int, end int) {
	if start >= end {
		return
	}
	i := start
	j := end
	base := nums[end]
	for i < j {
		for i < j && nums[i] <= base {
			i++
		}
		nums[j] = nums[i]
		for i < j && nums[j] >= base {
			j--
		}
		nums[i] = nums[j]
	}
	nums[i] = base
	quickSortHelper(nums, start, i-1)
	quickSortHelper(nums, j+1, end)
}

func quickSort(nums []int) {
	quickSortHelper(nums, 0, len(nums)-1)
}

func main() {
	nums := []int{49, 38, 65, 97, 76, 13, 27, 49}
	quickSort(nums)
	fmt.Println(nums)
}
